<?xml version="1.0" ?>
<!--This does not replace URDF, and is not an extension of URDF.
    This is a format for representing semantic information about the robot structure.
    A URDF file must exist for this robot as well, where the joints and the links that are referenced are defined
-->
<robot name="mue"
       xmlns:xacro="http://www.ros.org/wiki/xacro">
    <xacro:arg name="tf_prefix" default=""/>
    <!--GROUPS: Representation of a set of joints and links. This can be useful for specifying DOF to plan for, defining arms, end effectors, etc-->
    <!--LINKS: When a link is specified, the parent joint of that link (if it exists) is automatically included-->
    <!--JOINTS: When a joint is specified, the child link of that joint (which will always exist) is automatically included-->
    <!--CHAINS: When a chain is specified, all the links along the chain (including endpoints) are included in the group. Additionally, all the joints that are parents to included links are also included. This means that joints along the chain and the parent joint of the base link are included in the group-->
    <!--SUBGROUPS: Groups can also be formed by referencing to already defined group names-->
    <group name="ur5e_arm">
        <chain base_link="$(arg tf_prefix)base_link_ur5e" tip_link="$(arg tf_prefix)charger_base_link" />
    </group>
    <group name="endeffector_group">
        <chain base_link="$(arg tf_prefix)charger_base_link" tip_link="$(arg tf_prefix)charger_link" />
    </group>
    <!--GROUP STATES: Purpose: Define a named state for a particular group, in terms of joint values. This is useful to define states like 'folded arms'-->
    <group_state name="all_zeros" group="ur5e_arm">
        <joint name="$(arg tf_prefix)elbow_joint" value="0" />
        <joint name="$(arg tf_prefix)shoulder_lift_joint" value="0" />
        <joint name="$(arg tf_prefix)shoulder_pan_joint" value="0" />
        <joint name="$(arg tf_prefix)wrist_1_joint" value="0" />
        <joint name="$(arg tf_prefix)wrist_2_joint" value="0" />
        <joint name="$(arg tf_prefix)wrist_3_joint" value="0" />
        <!-- added joint -->
        <!-- <joint name="$(arg tf_prefix)ft_sensor_joint" value="0" /> -->
    </group_state>
    <group_state name="up" group="ur5e_arm">
        <joint name="$(arg tf_prefix)elbow_joint" value="0" />
        <joint name="$(arg tf_prefix)shoulder_lift_joint" value="-1.5708" />
        <joint name="$(arg tf_prefix)shoulder_pan_joint" value="0" />
        <joint name="$(arg tf_prefix)wrist_1_joint" value="1.5708" />
        <joint name="$(arg tf_prefix)wrist_2_joint" value="0" />
        <joint name="$(arg tf_prefix)wrist_3_joint" value="0" />
        <!--added ft sensor joint-->
        <!-- <joint name="$(arg tf_prefix)ft_sensor_joint" value="0" /> -->
    </group_state>
    <group_state name="test" group="ur5e_arm">
        <joint name="$(arg tf_prefix)elbow_joint" value="2.3" />
        <joint name="$(arg tf_prefix)shoulder_lift_joint" value="-1.3" />
        <joint name="$(arg tf_prefix)shoulder_pan_joint" value="0" />
        <joint name="$(arg tf_prefix)wrist_1_joint" value="-1" />
        <joint name="$(arg tf_prefix)wrist_2_joint" value="1.5708" />
        <joint name="$(arg tf_prefix)wrist_3_joint" value="0" />
        <!--added ft sensor joint-->
        <!-- <joint name="$(arg tf_prefix)ft_sensor_joint" value="0" /> -->
    </group_state>
    <!--END EFFECTOR: Purpose: Represent information about an end effector.-->
    <end_effector name="endeffector" parent_link="$(arg tf_prefix)wrist_3_link" group="endeffector_group" parent_group="ur5e_arm" />
    <!--VIRTUAL JOINT: Purpose: this element defines a virtual joint between a robot link and an external frame of reference (considered fixed with respect to the robot)-->
    <virtual_joint name="fixed_base" type="fixed" parent_frame="$(arg tf_prefix)base_link_ur5e" child_link="world" />
    <!--PASSIVE JOINT: Purpose: this element is used to mark joints that are not actuated-->
    <passive_joint name="$(arg tf_prefix)shoulder_pan_joint" />
    <passive_joint name="$(arg tf_prefix)shoulder_lift_joint" />
    <passive_joint name="$(arg tf_prefix)elbow_joint" />
    <passive_joint name="$(arg tf_prefix)wrist_1_joint" />
    <passive_joint name="$(arg tf_prefix)wrist_2_joint" />
    <passive_joint name="$(arg tf_prefix)wrist_3_joint" />
    <!-- <passive_joint name="$(arg tf_prefix)ft_sensor_joint" /> -->
    <!--DISABLE COLLISIONS: By default it is assumed that any link of the robot could potentially come into collision with any other link in the robot. This tag disables collision checking between a specified pair of links. -->
    <disable_collisions link1="$(arg tf_prefix)base_link_ur5e" link2="$(arg tf_prefix)shoulder_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)camera_base_link" link2="camera_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)camera_base_link" link2="$(arg tf_prefix)charger_base_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)camera_base_link" link2="$(arg tf_prefix)charger_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)camera_base_link" link2="$(arg tf_prefix)wrist_1_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)camera_base_link" link2="$(arg tf_prefix)wrist_2_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)camera_base_link" link2="$(arg tf_prefix)wrist_3_link" reason="Never" />
    <disable_collisions link1="camera_link" link2="$(arg tf_prefix)charger_base_link" reason="Never" />
    <disable_collisions link1="camera_link" link2="$(arg tf_prefix)charger_link" reason="Never" />
    <disable_collisions link1="camera_link" link2="$(arg tf_prefix)wrist_1_link" reason="Never" />
    <disable_collisions link1="camera_link" link2="$(arg tf_prefix)wrist_2_link" reason="Never" />
    <disable_collisions link1="camera_link" link2="$(arg tf_prefix)wrist_3_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)charger_base_link" link2="$(arg tf_prefix)charger_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)charger_base_link" link2="$(arg tf_prefix)wrist_1_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)charger_base_link" link2="$(arg tf_prefix)wrist_2_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)charger_base_link" link2="$(arg tf_prefix)wrist_3_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)charger_link" link2="$(arg tf_prefix)wrist_1_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)charger_link" link2="$(arg tf_prefix)wrist_2_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)charger_link" link2="$(arg tf_prefix)wrist_3_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)forearm_link" link2="$(arg tf_prefix)upper_arm_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)forearm_link" link2="$(arg tf_prefix)wrist_1_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)forearm_link" link2="$(arg tf_prefix)wrist_2_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)shoulder_link" link2="$(arg tf_prefix)upper_arm_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)wrist_1_link" link2="$(arg tf_prefix)wrist_2_link" reason="Adjacent" />
    <disable_collisions link1="$(arg tf_prefix)wrist_1_link" link2="$(arg tf_prefix)wrist_3_link" reason="Never" />
    <disable_collisions link1="$(arg tf_prefix)wrist_2_link" link2="$(arg tf_prefix)wrist_3_link" reason="Adjacent" />
</robot>
